File: app/src/main/java/com/github/premnirmal/ticker/portfolio/RearrangeActivity.java
package com.github.premnirmal.ticker.portfolio;

import android.graphics.Color;
import android.graphics.drawable.ColorDrawable;
import android.os.Bundle;
import android.support.v7.app.ActionBar;
import android.view.View;
import android.widget.AdapterView;

import com.github.premnirmal.ticker.BaseActivity;
import com.github.premnirmal.ticker.Injector;
import com.github.premnirmal.ticker.model.IStocksProvider;
import com.github.premnirmal.tickerwidget.R;
import com.nhaarman.listviewanimations.itemmanipulation.DynamicListView;

import javax.inject.Inject;

/**
 * Created by premnirmal on 3/7/15.
 */
public class RearrangeActivity extends BaseActivity {

    @Inject
    IStocksProvider stocksProvider;

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Injector.inject(this);
        final ActionBar actionBar = getSupportActionBar();
        actionBar.setDisplayHomeAsUpEnabled(true);
        setTitle(R.string.rearrange);
        final DynamicListView listView = new DynamicListView(this);
        final int spacing = getResources().getDimensionPixelSize(R.dimen.list_spacing);
        listView.setPadding(spacing,spacing,spacing,0);
        listView.setDivider(new ColorDrawable(Color.TRANSPARENT));
        listView.setDividerHeight(spacing);
        final RearrangeAdapter adapter = new RearrangeAdapter(stocksProvider);
        listView.setAdapter(adapter);
        listView.enableDragAndDrop();
        listView.setOnItemLongClickListener(new AdapterView.OnItemLongClickListener() {
            @Override
            public boolean onItemLongClick(AdapterView<?> parent, View view, int position, long id) {
                listView.startDragging(position);
                return true;
            }
        });
        setContentView(listView);
        showDialog(getString(R.string.drag_instructions));
    }

}

File: app/src/main/kotlin/com/github/premnirmal/ticker/portfolio/drag_drop/RearrangeActivity.kt
package com.github.premnirmal.ticker.portfolio.drag_drop

import android.content.SharedPreferences
import android.os.Bundle
import android.support.v7.widget.GridLayoutManager
import android.support.v7.widget.RecyclerView
import android.support.v7.widget.helper.ItemTouchHelper
import com.github.premnirmal.ticker.BaseActivity
import com.github.premnirmal.ticker.Injector
import com.github.premnirmal.ticker.Tools
import com.github.premnirmal.ticker.model.IStocksProvider
import com.github.premnirmal.ticker.ui.SpacingDecoration
import com.github.premnirmal.tickerwidget.R
import kotlinx.android.synthetic.main.activity_rearrange.*
import javax.inject.Inject

/**
 * Created by premnirmal on 2/29/16.
 */
class RearrangeActivity : BaseActivity(), OnStartDragListener {

  private var itemTouchHelper: ItemTouchHelper? = null

  @Inject
  lateinit internal var stocksProvider: IStocksProvider

  @Inject
  lateinit internal var preferences: SharedPreferences

  public override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    Injector.inject(this)
    setContentView(R.layout.activity_rearrange)
    updateToolbar(toolbar)
    toolbar.setNavigationOnClickListener {
      finish()
    }
    recyclerView.setHasFixedSize(true)
    val spacing = resources.getDimensionPixelSize(R.dimen.list_spacing)
    recyclerView.setPadding(spacing, spacing, spacing, 0)
    recyclerView.addItemDecoration(SpacingDecoration(spacing))
    val adapter = RearrangeAdapter(stocksProvider, this)
    recyclerView.adapter = adapter
    val layoutManager: RecyclerView.LayoutManager = GridLayoutManager(this, 2)
    recyclerView.layoutManager = layoutManager

    val callback: ItemTouchHelper.Callback = SimpleItemTouchHelperCallback(adapter)
    itemTouchHelper = ItemTouchHelper(callback)
    itemTouchHelper?.attachToRecyclerView(recyclerView)
    if (Tools.autoSortEnabled()) {
      preferences.edit().putBoolean(Tools.SETTING_AUTOSORT, false).apply()
      showDialog(getString(R.string.autosort_disabled))
    }
  }

  override fun onStartDrag(viewHolder: RecyclerView.ViewHolder) {
    itemTouchHelper?.startDrag(viewHolder)
  }
}